packages/grid/_modules_/grid/components/columnSelection/GridHeaderCheckbox.tsx

```diff
@@ -4,21 +4,22 @@ import {
   GRID_SELECTION_CHANGE,
 } from '../../constants/eventsConstants';
 import { useGridSelector } from '../../hooks/features/core/useGridSelector';
+import { gridPaginatedVisibleSortedGridRowIdsSelector } from '../../hooks/features/pagination/gridPaginationSelector';
 import { visibleSortedGridRowIdsSelector } from '../../hooks/features/filter/gridFilterSelector';
 import { gridTabIndexColumnHeaderSelector } from '../../hooks/features/focus/gridFocusStateSelector';
 import { gridRowCountSelector } from '../../hooks/features/rows/gridRowsSelector';
 import { selectedGridRowsCountSelector } from '../../hooks/features/selection/gridSelectionSelector';
 import { GridColumnHeaderParams } from '../../models/params/gridColumnHeaderParams';
 import { isNavigationKey, isSpaceKey } from '../../utils/keyboardUtils';
 import { useGridApiContext } from '../../hooks/root/useGridApiContext';
+import { optionsSelector } from '../../hooks/utils/optionsSelector';
 
 export const GridHeaderCheckbox = React.forwardRef<HTMLInputElement, GridColumnHeaderParams>(
   function GridHeaderCheckbox(props, ref) {
     const [, forceUpdate] = React.useState(false);
     const apiRef = useGridApiContext();
-    const visibleRowIds = useGridSelector(apiRef, visibleSortedGridRowIdsSelector);
+    const { checkboxSelectionVisibleOnly } = useGridSelector(apiRef, optionsSelector);
```

Comment:
https://github.com/mui-org/material-ui-x/pull/1952#discussion_r656968843

Like @oliviertassinari , I'd prefer to avoid destructuring for widely used objects like `options`. 
It make it easier to search for usage across the codebase.

```suggestion
    const options = useGridSelector(apiRef, optionsSelector);
```